### PR TITLE
Add dedicated audit logging setup for managed gardener shoot 

### LINKED ISSUE #1414: KIM claims SiemApproved AuditLog CRs for Runtimes and marks them Orphaned on deletion
## Description

KIM (Kyma Infrastructure Manager) must take ownership of the `AuditLog` CR lifecycle at two critical points:

1. **Assignment** — when a Runtime with `auditLogAccessEnabled: true` is provisioned, KIM picks a free `AuditLog` CR (state `SiemApproved`, not yet assigned) and records the claim in both the `Runtime` CR and the `AuditLog` CR.
2. **Orphaning** — when a Runtime is deleted, KIM marks the associated `AuditLog` CR as `Orphaned` so the Auditlog Controller (KALC) can start the retention countdown and eventually clean up the BTP resources.

This issue is the KIM-side counterpart to the KALC controller implemented in [kyma-auditlog-manager](https://github.tools.sap/kyma/kyma-auditlog-manager/issues/9). The `AuditLog` CR lifecycle states are: `Pending` → `RegistrationReady` → `SiemApproved` → `Assigned` → `Orphaned`.

## Prerequisites

- #1408 — `Runtime` CR must have `spec.auditLogAccessEnabled` and `spec.auditLogCR` fields

### 1. Claim a free AuditLog CR (Assignment)

When KIM reconciles a `Runtime` CR where `spec.auditLogAccessEnabled == true` and `spec.auditLogCR` is empty:

1. List all `AuditLog` CRs in the KCP namespace.
2. Select one where `status.state == SiemApproved` and `status.assignedToRuntime` is empty.
3. If none are available, requeue and wait (do not block Runtime provisioning — log a warning and continue without audit log assignment; retry on next reconciliation).
4. Write the selected `AuditLog` CR name into `spec.auditLogCR` on the `Runtime` CR.
5. Patch `status.assignedToRuntime` on the `AuditLog` CR with the Runtime's name/ID and advance its `status.state` to `Assigned`.
6. KALC will react to the state change and synchronise the customer-access service binding to the SKR (out of scope for this issue).

### 2. Mark AuditLog CR as Orphaned (Deletion)

When KIM reconciles the deletion of a `Runtime` CR where `spec.auditLogCR` is set:

1. Look up the referenced `AuditLog` CR by name from `spec.auditLogCR`.
2. If found and `status.state == Assigned`, patch its `status.state` to `Orphaned` and set `status.orphanedAt` to the current timestamp.
3. Clear `spec.auditLogCR` on the `Runtime` CR as part of the deletion reconciliation.
4. KALC will handle the retention countdown and BTP resource cleanup via the Orphaned state (out of scope for this issue).

## Non-Functional Requirements

- **Idempotent**: if `spec.auditLogCR` is already set on the `Runtime` CR, skip the selection step and verify the referenced `AuditLog` CR is in state `Assigned`.
- **Safe concurrent access**: use optimistic locking (resourceVersion) when patching the `AuditLog` CR to avoid double-claiming the same instance by concurrent reconciliations.
- **No BTP calls**: KIM only manages CR state — all BTP provisioning and credential handling is done by KALC.
- **Graceful degradation**: if no `SiemApproved` instance is available at assignment time, Runtime provisioning must not be blocked; the reconciler should retry without failing the Runtime.

## Acceptance Criteria

- [ ] When a `Runtime` CR with `spec.auditLogAccessEnabled: true` is reconciled, KIM selects a free `AuditLog` CR (`status.state == SiemApproved`, `status.assignedToRuntime` empty) and writes its name into `spec.auditLogCR`
- [ ] After selection, the `AuditLog` CR has `status.state == Assigned` and `status.assignedToRuntime` set to the Runtime's identifier
- [ ] Concurrent reconciliations do not double-claim the same `AuditLog` CR (optimistic locking / conflict retry)
- [ ] If no `SiemApproved` `AuditLog` CR is available, KIM logs a ERROR and requeues the `Runtime` CR. It is not allowed to switch into `Ready` state - the Runtime-CR is not correctly configured and has to stay in status `Pending`.
    - [ ] As soon as a `SiemApproved`Auditlog CR becomes available, KIM will in the next reconciliation run of a Runtime update the Auditlog-configuration in the Shoot-Spec  and `Runtime`-CR can reach the `Ready`state.
- [ ] When a `Runtime` CR is deleted and `spec.auditLogCR` is set, the referenced `AuditLog` CR is patched to `status.state == Orphaned` with `status.orphanedAt` timestamp
- [ ] If the referenced `AuditLog` CR is not found at deletion time, the error is logged and deletion proceeds (best-effort cleanup)
- [ ] The reconciler is idempotent: re-running with an already-assigned `AuditLog` CR produces no duplicate patches
- [ ] Unit tests cover: successful claim, no available instances (requeue path), successful orphan marking, missing `AuditLog` CR at deletion

## Context & Related Issues

- kyma-auditlog-manager controller & CRD: https://github.tools.sap/kyma/kyma-auditlog-manager/issues/9
- kyma-auditlog-manager pool manager CronJob: https://github.tools.sap/kyma/kyma-auditlog-manager/issues/6
- Runtime CR extension (prerequisite): #1408
- Epic: https://github.tools.sap/kyma/backlog/issues/7888

### PR DESCRIPTION
**Pull request by koala7659**
## Description

Implements dedicated BTP audit logging infrastructure integration with Kyma Runtime provisioning using a **validation-first, migrate-last provisioning model** with two-phase resource claiming.

Implements: https://github.com/kyma-project/kyma-infrastructure-manager/issues/1414

## Key Documents for Review

- 📋 **ADR Document**: [docs/adr/004-dedicated-audit-logging.md](https://github.com/koala7659/infrastructure-manager/blob/use-dedicated-auditlog-config/docs/adr/004-dedicated-audit-logging.md)
- 🔧 **Implementation Details**: [docs/adr/004-dedicated-audit-logging-implementation.md](https://github.com/koala7659/infrastructure-manager/blob/use-dedicated-auditlog-config/docs/adr/004-dedicated-audit-logging-implementation.md)

### Two-Phase Claiming Mechanism

**Phase 1: Reserve (Light Lock)** - During `sFnCreateShoot`
- Validates dedicated AuditLog availability in pool
- Adds reservation labels to prevent race conditions
- Creates shoot with shared audit logging

**Phase 2: Claim (Heavy Lock)** - During `sFnMigrateToDedicatedAuditLog` (new final FSM state)
- Upgrades reservation to claim via `spec.assignedToRuntimeID`
- Patches shoot with dedicated audit log configuration
- Two-reconciliation pattern: patch → requeue → verify → complete

### Provisioning Flow

```
sFnCreateShoot (validates + reserves)
  ↓ (shoot created with shared logging)
[... standard provisioning ...]
  ↓
sFnMigrateToDedicatedAuditLog (claims + patches - FINAL STEP)
  ↓
Complete
```

## Implementation

**Core Components:**
- `pkg/auditlog/provider.go` - Data provider interface with Reserve/Claim/Release
- `pkg/auditlog/dedicated_client.go` - Two-phase claim implementation
- `internal/controller/runtime/fsm/runtime_fsm_migrate_dedicated_auditlog.go` - New FSM state
- `internal/controller/runtime/fsm/runtime_fsm_create_shoot.go` - Validation + reservation

**Configuration:**
- Feature flag: `auditlog.enable-dedicated-auditlog-stack: true`
- Runtime opt-in: `spec.shoot.auditLogAccessEnabled: true`

**Key Features:**
- ✅ Fail-fast validation before shoot creation
- ✅ Race condition prevention via reservation labels
- ✅ Idempotent operations (safe retries)
- ✅ RegistrationReady + SiemApproved state support
- ✅ Kubernetes optimistic concurrency handling

**Testing:**
- 37 comprehensive unit tests covering all edge cases
- `pkg/auditlog/provider_test.go` - 14 tests
- `pkg/auditlog/dedicated_client_test.go` - 23 tests

## Consequences

**Positive:**
- No wasted pool resources (claim only after full provisioning)
- Race conditions prevented (labels lock during ~5-15min provisioning)
- Observable pool state via label selectors
- No KALM CRD changes required

**Negative:**
- Brief shared logging period during provisioning
- Manual cleanup of stale reservations (MVP - automation in future)

## Breaking Changes

None. Feature is opt-in via feature flag + runtime field.

## Next Steps

See https://github.com/kyma-project/kyma-infrastructure-manager/issues/1496
- Copy read credentials to runtime cluster

### CODE REVIEW

File: api/v1/runtime_types.go [hunk 1]
```diff
@@ -94,7 +95,9 @@ const (
 	ConditionReasonGardenerError           = RuntimeConditionReason("GardenerErr")
 	ConditionReasonKubernetesAPIErr        = RuntimeConditionReason("KubernetesErr")
 
-	ConditionReasonAuditLogError = RuntimeConditionReason("AuditLogErr")
+	ConditionReasonAuditLogError            = RuntimeConditionReason("AuditLogErr")
```
Comment (szwedm) [inline api/v1/runtime_types.go:98]: nit: naming consistency `Err` vs `Error`

File: config/crd/bases/infrastructuremanager.kyma-project.io_runtimes.yaml [hunk 2]
```diff
@@ -555,6 +555,33 @@ spec:
                                   required:
                                   - provider
                                   type: object
+                                exposure:
```
Comment (szwedm) [inline config/crd/bases/infrastructuremanager.kyma-project.io_runtimes.yaml:558]: Looks like a accidental addition. Is the `exposure` property related to Audit Logs?

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 3]
```diff
@@ -56,18 +56,36 @@ func sFnApplyClusterRoleBindings(ctx context.Context, m *fsm, s *systemState) (s
...
+	m.log.Info("Finished configuring shoot")
+
+	// Only proceed to audit log migration if both flags are enabled
+	if m.DedicatedAuditLoggingEnabled &&
+		s.instance.Spec.AuditLogAccessEnabled != nil &&
+		*s.instance.Spec.AuditLogAccessEnabled {
+
+		s.instance.UpdateStatePending(
+			imv1.ConditionTypeRuntimeConfigured,
+			imv1.ConditionReasonAdministratorsConfigured,
```
Comment (szwedm) [inline internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:68]: Are conditions correct for the Audit Log enablement?

File: config/crd/bases/infrastructuremanager.kyma-project.io_runtimes.yaml [hunk 4]
```diff
@@ -909,7 +936,7 @@ spec:
                                       properties:
                                         swapBehavior:
                                           description: |-
-                                            SwapBehavior configures swap memory available to container workloads. May be one of {"LimitedSwap", "UnlimitedSwap"}
+                                            SwapBehavior configures swap memory available to container workloads. May be one of {"NoSwap", "LimitedSwap"}
```
Comment (Disper) [inline config/crd/bases/infrastructuremanager.kyma-project.io_runtimes.yaml:939]: This field is optional and I don't see that we're using it in KIM. I believe it's fine to add it here and align with gardener shoot_type.

File: config/crd/bases/infrastructuremanager.kyma-project.io_runtimes.yaml (hunk 2 above)
Comment (Disper) [reply config/crd/bases/infrastructuremanager.kyma-project.io_runtimes.yaml:558]: We use the gardener type for controlPlane https://github.com/kyma-project/kyma-infrastructure-manager/blob/3debcdf4ace3abf2391719bd9fcad288bc7a59a6/api/v1/runtime_types.go#L200 so we anyway use this API. I think we could keep this CRD update for the sake of the consistency.

Comment (koala7659) [reply config/crd/bases/infrastructuremanager.kyma-project.io_runtimes.yaml:558]: The answer is interesting - this is Gardener API update introduced few months ago. Should not be a problem

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go (hunk 3 above)
Comment (koala7659) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:68]: It is only to show that current state "Cluster admin configuration" is completed, but we are still in the Pending state

File: internal/controller/runtime/fsm/runtime_fsm_create_shoot.go [hunk 5]
```diff
@@ -50,6 +51,38 @@ func sFnCreateShoot(ctx context.Context, m *fsm, s *systemState) (stateFn, *ctrl
...
+			"runtimeID", s.instance.GetName(),
+			"region", s.instance.Spec.Shoot.Region)
+
+		// Phase 1: Reserve an AuditLogCR by adding labels (light lock)
+		// The region is the hyperscaler region (e.g., eu-central-1) that must match
+		// one of the AuditLogCR's spec.regions entries
+		err := m.AuditLogDataProvider.ReserveAuditLog(
+			ctx,
+			s.instance.Spec.Shoot.Region,
+			s.instance.Labels[imv1.LabelKymaRuntimeID],
```
[the remark this answers is not in the document]
Comment (koala7659) [reply internal/controller/runtime/fsm/runtime_fsm_create_shoot.go]: Will be fixed

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 6]
```diff
@@ -56,18 +56,36 @@ func sFnApplyClusterRoleBindings(ctx context.Context, m *fsm, s *systemState) (s
...
+			metav1.ConditionTrue,
+			"Cluster admin configuration complete",
+		)
```
[the remark this answers is not in the document]
Comment (koala7659) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:71]: This just set condition status as Cluster admin configuration complete is completed

File: pkg/auditlog/v1beta1/auditlog_types.go
Comment (szwedm) [inline pkg/auditlog/v1beta1/auditlog_types.go:1]: We need to think how to automatically sync AuditLog structs between the repositories

File: pkg/auditlog/dedicated_client_test.go [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+			WithRuntimeObjects(&auditLog).
+			Build()
+
+		provider := &DefaultDataProvider{
+			client: fakeClient,
+			logger: logger,
+		}
+
+		err := provider.reserveAuditLogCR(context.Background(), "test-runtime", "eu-central-1")
+		require.NoError(t, err)
```
Comment (szwedm) [inline pkg/auditlog/dedicated_client_test.go]: Misleading test name, assertion is for error from `reserveAuditLogCR` method

File: pkg/auditlog/dedicated_client.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+func (p *DefaultDataProvider) releaseAuditLogCR(ctx context.Context, auditLog *auditlogv1.AuditLog) error {
+	if auditLog.Status.State != auditlogv1.StateAssigned {
+		p.logger.Info("AuditLogCR is not in Assigned state, skipping release",
+			"name", auditLog.Name, "state", auditLog.Status.State)
+		return nil
+	}
+
+	// Mark as orphaned - KALM will handle the retention period and cleanup
+	auditLog.Spec.Orphaned = true
+	if err := p.client.Update(ctx, auditLog); err != nil {
```
Comment (szwedm) [inline pkg/auditlog/dedicated_client.go:221]: Does setting spec.Orphaned to true implicitly sets status.OrphanedAt? OrphanedAt is required for establishing the retention time before cleanup in KALM

File: pkg/auditlog/provider_test.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+		fakeClient := fake.NewClientBuilder().
+			WithScheme(scheme).
+			WithRuntimeObjects(&auditLog).
+			Build()
+
+		provider := NewDataProvider(fakeClient, nil, logger)
+
+		err := provider.ReserveAuditLog(context.Background(), "eu-central-1", "test-runtime")
+		require.NoError(t, err)
+	})
```
Comment (szwedm) [inline pkg/auditlog/provider_test.go]: Misleading test description, the test asserts error occurrence

File: pkg/auditlog/dedicated_client.go (hunk 8 above)
Comment (koala7659) [reply pkg/auditlog/dedicated_client.go:221]: The value of `OrphanedAt` in status is calculated by KALM controller and set to the current time.

File: pkg/auditlog/dedicated_client_test.go (hunk 7 above)
Comment (koala7659) [reply pkg/auditlog/dedicated_client_test.go]: Fixing - will change it to "succeeds when CR already reserved for same runtime (idempotent)"

File: pkg/auditlog/provider_test.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+			WithScheme(scheme).
+			WithRuntimeObjects(&auditLog).
+			WithIndex(&auditlogv1.AuditLog{}, "spec.assignedToRuntimeID", indexByAssignedRuntimeID).
+			Build()
+
+		provider := NewDataProvider(fakeClient, nil, logger)
+
+		data, err := provider.GetDedicatedAuditLogData(context.Background(), "test-runtime", false)
+		require.NoError(t, err)
+		assert.Equal(t, "dedicated-tenant", data.TenantID)
```
Comment (szwedm) [inline pkg/auditlog/provider_test.go:188]: Looks like the test is missing an assertion for AssignedToRuntimeID field to determine if the AuditLog has been claimed or not

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go (hunk 3 above)
Comment (szwedm) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:68]: I see, so the state is about what's been completed so far.

File: pkg/auditlog/provider_test.go (hunk 10 above)
Comment (koala7659) [reply pkg/auditlog/provider_test.go:188]: Fixed

File: pkg/auditlog/provider_test.go (hunk 9 above)
Comment (koala7659) [reply pkg/auditlog/provider_test.go]: Changed to "successfully reserves available CR"

**Review by m00g3n — APPROVED** [review]
LGTM

**Review by IwonaLanger — APPROVED** [review]
